qiime2/sdk/usage.py

```diff
@@ -8,60 +8,194 @@
+                not callable(assert_has_line_matching):
+            raise TypeError('Value for `assert_has_line_matching` should be a '
+                            '`callable`.')
 
-        self.plugin_id = plugin_id
-        self.action_id = action_id
-        self._plugin_manager = sdk.PluginManager()
+        self.ref = ref
+        self._result = value
+        # TODO: Should we put a guard somewhere for acceptable sources?
+        self._source = source
+        self._assert_has_line_matching_ = assert_has_line_matching
 
     def __repr__(self):
-        return 'UsageAction(plugin_id=%r, action_id=%r)' %\
-            (self.plugin_id, self.action_id)
+        return 'ScopeRecord<ref=%s, result=%r, source=%s>' % (self.ref,
+                                                              self.result,
+                                                              self.source)
 
-    def get_action(self):
-        plugin = self._plugin_manager.get_plugin(id=self.plugin_id)
-        # TODO: should this validation be pushed up into
-        # plugin.py or action.py?
-        try:
-            action_f = plugin.actions[self.action_id]
-        except KeyError:
-            raise KeyError('No action currently registered with '
-                           'id: "%s".' % (self.action_id,))
-        return (action_f, action_f.signature)
+    @property
+    def result(self) -> ActionData:
+        """
+        Union[sdk.Artifact, sdk.Visualization, metadata.Metadata]: The value
+        referred to by `self.ref`
+        """
+        return self._result
 
-    def validate(self, inputs, outputs):
-        if not isinstance(inputs, UsageInputs):
-            raise TypeError('Must provide an instance of UsageInputs.')
-        if not isinstance(outputs, UsageOutputNames):
-            raise TypeError('Must provide an instance of UsageOutputNames.')
+    @property
+    def source(self) -> str:
+        """
+        str: The Usage method called to initialize example data.  This is used
+        by Usage drivers to render Usage examples.
+        """
+        return self._source
 
-        _, sig = self.get_action()
+    def assert_has_line_matching(
+            self, label: str, path: str, expression: str
+    ) -> None:
+        """
+        Verify that the file at `path` contains a line matching `expression`
+
+        Parameters
+        ----------
+        label : str
+            A label for describing this assertion
+        path : str
+            Path to example data file
+        expression : str
+            a regex pattern to be passed as the first argument to `re.search`
+
+        Returns
+        -------
+        None
+
+        Raises
+        ______
+        AssertionError
+            If `expression` is not found in `path`
+
+        See Also
+        --------
+        See `ExecutionUsage` for an example implementation
+        """
+        # TODO: mypy reports "None" not callable. Should we refactor?
```

Comment:
Schwat?

Reply:
Sorry, that was not at all clear without additional context.

There's a guard in the `ScopeRecord` constructor which makes sure that `assert_has_line_matching` is either a callable or `None`.  But if `assert_has_line_matching` *was* `None`, the line below would break since `None` is not callable.  Which mypy doesn't like.

```python
return self._assert_has_line_matching_(self.ref, label, path, expression)
````

So my question is, does the mypy error indicate that a slight refactor is desirable?  How much do we care about what mypy thinks?

Reply:
Right now its fine, because if an interface chooses not to implement `assert_has_line_matching`, those statements in the usage examples are skipped.